File: calendar/base/backend/icaljs/calRecurrenceRule.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

Components.utils.import("resource://calendar/modules/ical.js");
Components.utils.import("resource://calendar/modules/calUtils.jsm");
Components.utils.import("resource://gre/modules/XPCOMUtils.jsm");

function calRecurrenceRule(innerObject) {
    this.innerObject = innerObject || new ICAL.Recur();
    this.wrappedJSObject = this;
}

var calRecurrenceRuleInterfaces = [Components.interfaces.calIRecurrenceRule];
var calRecurrenceRuleClassID = Components.ID("{df19281a-5389-4146-b941-798cb93a7f0d}");
calRecurrenceRule.prototype = {
    QueryInterface: XPCOMUtils.generateQI(calRecurrenceRuleInterfaces),
    classID: calRecurrenceRuleClassID,
    classInfo: XPCOMUtils.generateCI({
        contractID: "@mozilla.org/calendar/recurrence-rule;1",
        classDescription: "Calendar Recurrence Rule",
        classID: calRecurrenceRuleClassID,
        interfaces: calRecurrenceRuleInterfaces
    }),

    innerObject: null,

    isMutable: true,
    makeImmutable: function() { this.isMutable = false; },
    clone: function() { return new calRecurrenceRule(new ICAL.Recur(this.innerObject)); },

    isNegative: false, // We don't support EXRULE anymore
    get isFinite() { return this.innerObject.isFinite(); },

    getNextOccurrence: function(aStartTime, aRecId) {
        aStartTime = unwrapSingle(ICAL.Time, aStartTime);
        aRecId = unwrapSingle(ICAL.Time, aRecId);
        return wrapGetter(calDateTime, this.innerObject.getNextOccurrence(aStartTime, aRecId));
    },

    getOccurrences: function(aStartTime, aRangeStart, aRangeEnd, aMaxCount, aCount) {
        aStartTime = unwrapSingle(ICAL.Time, aStartTime);
        aRangeStart = unwrapSingle(ICAL.Time, aRangeStart);
        aRangeEnd = unwrapSingle(ICAL.Time, aRangeEnd);

        if (!aMaxCount && !aRangeEnd && this.count == 0 && this.until == null) {
            throw Components.results.NS_ERROR_INVALID_ARG;
        }

        let occurrences = [];
        let rangeStart = aRangeStart.clone();
        rangeStart.isDate = false;

        let dtend = null;

        if (aRangeEnd) {
            dtend = aRangeEnd.clone();
            dtend.isDate = false;

            // If the start of the recurrence is past the end, we have no dates
            if (aStartTime.compare(dtend) >= 0) {
                aCount.value = 0;
                return [];
            }
        }

        let iter = this.innerObject.iterator(aStartTime);

        for (let next = iter.next(); next ; next = iter.next()) {
            let dtNext  = next.clone();
            dtNext.isDate = false;

            if (dtNext.compare(rangeStart) < 0) {
                continue;
            }

            if (dtend && dtNext.compare(dtend) >= 0) {
                break;
            }

            next = next.clone();

            if (aStartTime.zone) {
                next.zone = aStartTime.zone;
            }

            occurrences.push(new calDateTime(next));

            if (aMaxCount && aMaxCount >= occurrences.length) {
                break;
            }
        }

        aCount.value = occurrences.length;
        return occurrences;
    },

    get icalString() { return "RRULE:" + this.innerObject.toString() + ICAL.newLineChar; },
    set icalString(val) { this.innerObject = ICAL.Recur.fromString(val.replace(/^RRULE:/i, "")); },

    get icalProperty() {
        let prop = new ICAL.Property("rrule");
        prop.setValue(this.innerObject);
        return new calIcalProperty(prop);
    },
    set icalProperty(val) { unwrapSetter(ICAL.Property, val, function(val) {
        this.innerObject = val.getFirstValue();
    }, this); },

    get type() { return this.innerObject.freq; },
    set type(val) { this.innerObject.freq = val; },

    get interval() { return this.innerObject.interval; },
    set interval(val) { this.innerObject.interval = val; },

    get count() {
        if (!this.isByCount) {
            throw Components.results.NS_ERROR_FAILURE;
        }
        return this.innerObject.count || -1;
    },
    set count(val) { this.innerObject.count = (val && val > 0 ? val : null); },

    get untilDate() {
        if (this.innerObject.until) {
            return new calDateTime(this.innerObject.until);
        } else {
            return null;
        }
    },
    set untilDate(val) { unwrapSetter(ICAL.Time, val, function(val) {
        if (val.timezone != ICAL.Timezone.utcTimezone &&
            val.timezone != ICAL.Timezone.localTimezone) {
            val = val.convertToZone(ICAL.Timezone.utcTimezone);
        }

        this.innerObject.until = val;
    }, this); },

    get isByCount() { return this.innerObject.isByCount(); },

    get weekStart() { return this.innerObject.wkst - 1; },
    set weekStart(val) { this.innerObject.wkst = val + 1; },

    getComponent: function(aType, aCount) {
        let values = this.innerObject.getComponent(aType);
        if (aType == "BYDAY") {
            // BYDAY values are alphanumeric: SU, MO, TU, etc..
            for (let i = 0; i < values.length; i++) {
                let match = /^([+-])?(5[0-3]|[1-4][0-9]|[1-9])?(SU|MO|TU|WE|TH|FR|SA)$/.exec(values[i]);
                if (!match) {
                    cal.ERROR("Malformed BYDAY rule\n" + cal.STACK(10));
                    return [];
                }
                values[i] = ICAL.Recur.icalDayToNumericDay(match[3]);
                if (match[2]) {
                    // match[2] is the week number for this value.
                    values[i] += 8 * match[2];
                }
                if (match[1] == '-') {
                    // Week numbers are counted back from the end of the period.
                    values[i] *= -1;
                }
            }
        }

        if (aCount) aCount.value = values.length;
        return values;
    },

    setComponent: function(aType, aCount, aValues) {
        let values = aValues;
        if (aType == "BYDAY") {
            // BYDAY values are alphanumeric: SU, MO, TU, etc..
            for (let i = 0; i < values.length; i++) {
                let absValue = Math.abs(values[i]);
                if (absValue > 7) {
                    let ordinal = Math.trunc(values[i] / 8);
                    let day = ICAL.Recur.numericDayToIcalDay(absValue % 8);
                    values[i] = ordinal + day;
                } else {
                    values[i] = ICAL.Recur.numericDayToIcalDay(values[i]);
                }
            }
        }
        this.innerObject.setComponent(aType, values);
    }
};

File: calendar/base/backend/libical/calRecurrenceRule.cpp
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "nsCOMArray.h"

#include "calRecurrenceRule.h"

#include "calDateTime.h"
#include "calIItemBase.h"
#include "calIEvent.h"

#include "calICSService.h"

#include "nsIClassInfoImpl.h"

#include <climits>

NS_IMPL_CLASSINFO(calRecurrenceRule, NULL, 0, CAL_RECURRENCERULE_CID)
NS_IMPL_ISUPPORTS_CI(calRecurrenceRule, calIRecurrenceItem, calIRecurrenceRule)

calRecurrenceRule::calRecurrenceRule()
    : mImmutable(false), mIsNegative(false), mIsByCount(false) {
  icalrecurrencetype_clear(&mIcalRecur);
}

NS_IMETHODIMP
calRecurrenceRule::GetIsMutable(bool* aResult) {
  NS_ENSURE_ARG_POINTER(aResult);
  *aResult = !mImmutable;
  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::MakeImmutable() {
  mImmutable = true;
  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::Clone(calIRecurrenceItem** aResult) {
  calRecurrenceRule* const crc = new calRecurrenceRule();
  CAL_ENSURE_MEMORY(crc);

  crc->mIsNegative = mIsNegative;
  crc->mIsByCount = mIsByCount;
  crc->mIcalRecur = mIcalRecur;

  NS_ADDREF(*aResult = crc);
  return NS_OK;
}

/* attribute boolean isNegative; */
NS_IMETHODIMP
calRecurrenceRule::GetIsNegative(bool* _retval) {
  NS_ENSURE_ARG_POINTER(_retval);
  *_retval = mIsNegative;
  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::SetIsNegative(bool aIsNegative) {
  if (mImmutable) return NS_ERROR_OBJECT_IS_IMMUTABLE;
  mIsNegative = aIsNegative;
  return NS_OK;
}

/* readonly attribute boolean isFinite; */
NS_IMETHODIMP
calRecurrenceRule::GetIsFinite(bool* _retval) {
  NS_ENSURE_ARG_POINTER(_retval);

  if ((mIsByCount && mIcalRecur.count == 0) ||
      (!mIsByCount && icaltime_is_null_time(mIcalRecur.until))) {
    *_retval = false;
  } else {
    *_retval = true;
  }
  return NS_OK;
}

/* attribute long type; */
NS_IMETHODIMP
calRecurrenceRule::GetType(nsACString& aType) {
  switch (mIcalRecur.freq) {
#define RECUR_HELPER(x)       \
  case ICAL_##x##_RECURRENCE: \
    aType.AssignLiteral(#x);  \
    break
    RECUR_HELPER(SECONDLY);
    RECUR_HELPER(MINUTELY);
    RECUR_HELPER(HOURLY);
    RECUR_HELPER(DAILY);
    RECUR_HELPER(WEEKLY);
    RECUR_HELPER(MONTHLY);
    RECUR_HELPER(YEARLY);
#undef RECUR_HELPER
    default:
      aType.AssignLiteral("");
  }

  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::SetType(const nsACString& aType) {
  if (mImmutable) return NS_ERROR_OBJECT_IS_IMMUTABLE;
#define RECUR_HELPER(x) \
  if (aType.EqualsLiteral(#x)) mIcalRecur.freq = ICAL_##x##_RECURRENCE
  RECUR_HELPER(SECONDLY);
  else RECUR_HELPER(MINUTELY);
  else RECUR_HELPER(HOURLY);
  else RECUR_HELPER(DAILY);
  else RECUR_HELPER(WEEKLY);
  else RECUR_HELPER(MONTHLY);
  else RECUR_HELPER(YEARLY);
#undef RECUR_HELPER
  else if (aType.IsEmpty() || aType.EqualsLiteral("")) mIcalRecur.freq =
      ICAL_NO_RECURRENCE;
  else return NS_ERROR_FAILURE;

  return NS_OK;
}

/* attribute long count; */
NS_IMETHODIMP
calRecurrenceRule::GetCount(int32_t* aRecurCount) {
  NS_ENSURE_ARG_POINTER(aRecurCount);

  if (!mIsByCount) return NS_ERROR_FAILURE;

  if (mIcalRecur.count == 0 && icaltime_is_null_time(mIcalRecur.until)) {
    *aRecurCount = -1;
  } else if (mIcalRecur.count) {
    *aRecurCount = mIcalRecur.count;
  } else {
    // count wasn't set, so we don't know
    return NS_ERROR_FAILURE;
  }

  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::SetCount(int32_t aRecurCount) {
  if (mImmutable) return NS_ERROR_OBJECT_IS_IMMUTABLE;
  if (aRecurCount != -1) {
    if (aRecurCount < 0 || aRecurCount > INT_MAX) return NS_ERROR_ILLEGAL_VALUE;
    mIcalRecur.count = static_cast<int>(aRecurCount);
    mIsByCount = true;
  } else {
    mIcalRecur.count = 0;
    mIsByCount = false;
  }

  mIcalRecur.until = icaltime_null_time();

  return NS_OK;
}

/* attribute calIDateTime untilDate; */
NS_IMETHODIMP
calRecurrenceRule::GetUntilDate(calIDateTime** aRecurEnd) {
  NS_ENSURE_ARG_POINTER(aRecurEnd);

  if (mIsByCount) return NS_ERROR_FAILURE;

  if (!icaltime_is_null_time(mIcalRecur.until)) {
    *aRecurEnd = new calDateTime(&mIcalRecur.until, nullptr);
    CAL_ENSURE_MEMORY(*aRecurEnd);
    NS_ADDREF(*aRecurEnd);
  } else {
    // infinite recurrence
    *aRecurEnd = nullptr;
  }
  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::SetUntilDate(calIDateTime* aRecurEnd) {
  if (mImmutable) return NS_ERROR_OBJECT_IS_IMMUTABLE;
  if (aRecurEnd) {
    nsresult rv;
    bool b;
    nsCOMPtr<calIDateTimeLibical> icaldt;
    nsCOMPtr<calITimezone> tz;
    aRecurEnd->GetTimezone(getter_AddRefs(tz));

    if (NS_SUCCEEDED(tz->GetIsUTC(&b)) && !b &&
        NS_SUCCEEDED(tz->GetIsFloating(&b)) && !b) {
      // convert to UTC:
      nsCOMPtr<calIDateTime> dt;
      nsCOMPtr<calITimezone> ctz = cal::UTC();
      aRecurEnd->GetInTimezone(ctz, getter_AddRefs(dt));
      icaldt = do_QueryInterface(dt, &rv);
    } else {
      icaldt = do_QueryInterface(aRecurEnd, &rv);
    }

    NS_ENSURE_SUCCESS(rv, rv);
    struct icaltimetype itt;
    icaldt->ToIcalTime(&itt);

    mIcalRecur.until = itt;
  } else {
    mIcalRecur.until = icaltime_null_time();
  }

  mIcalRecur.count = 0;

  mIsByCount = false;

  return NS_OK;
}

/* readonly attribute boolean isByCount; */
NS_IMETHODIMP
calRecurrenceRule::GetIsByCount(bool* aIsByCount) {
  *aIsByCount = mIsByCount;
  return NS_OK;
}

/* attribute long interval; */
NS_IMETHODIMP
calRecurrenceRule::GetInterval(int32_t* aInterval) {
  NS_ENSURE_ARG_POINTER(aInterval);
  *aInterval = mIcalRecur.interval;
  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::SetInterval(int32_t aInterval) {
  if (mImmutable) return NS_ERROR_OBJECT_IS_IMMUTABLE;
  if (aInterval < 0 || aInterval > SHRT_MAX) return NS_ERROR_ILLEGAL_VALUE;
  mIcalRecur.interval = static_cast<short>(aInterval);
  return NS_OK;
}

// Helper table to encode the size/location of the various arrays in the
// icalrecurrencetype struct.
static const struct {
  const char* name;
  size_t offset;
  size_t maxCount;
} recurrenceTable[] = {
    {"BYSECOND", offsetof(icalrecurrencetype, by_second), ICAL_BY_SECOND_SIZE},
    {"BYMINUTE", offsetof(icalrecurrencetype, by_minute), ICAL_BY_MINUTE_SIZE},
    {"BYHOUR", offsetof(icalrecurrencetype, by_hour), ICAL_BY_HOUR_SIZE},
    {"BYDAY", offsetof(icalrecurrencetype, by_day), ICAL_BY_DAY_SIZE},
    {"BYMONTHDAY", offsetof(icalrecurrencetype, by_month_day),
     ICAL_BY_MONTHDAY_SIZE},
    {"BYYEARDAY", offsetof(icalrecurrencetype, by_year_day),
     ICAL_BY_YEARDAY_SIZE},
    {"BYWEEKNO", offsetof(icalrecurrencetype, by_week_no), ICAL_BY_WEEKNO_SIZE},
    {"BYMONTH", offsetof(icalrecurrencetype, by_month), ICAL_BY_MONTH_SIZE},
    {"BYSETPOS", offsetof(icalrecurrencetype, by_set_pos), ICAL_BY_SETPOS_SIZE},
    {nullptr, 0, 0}};

NS_IMETHODIMP
calRecurrenceRule::GetComponent(const nsACString& aComponentType,
                                nsTArray<int16_t>& aValues) {
  aValues.ClearAndRetainStorage();
  // Look up the array for this component type.
  for (int i = 0; recurrenceTable[i].name; ++i) {
    auto const& row = recurrenceTable[i];
    if (aComponentType.EqualsASCII(row.name)) {
      // Found it.
      int16_t const* src = (int16_t*)((uint8_t*)&mIcalRecur + row.offset);
      size_t count;
      for (count = 0; count < row.maxCount; count++) {
        if (src[count] == ICAL_RECURRENCE_ARRAY_MAX) break;
      }
      aValues.ReplaceElementsAt(0, aValues.Length(), src, count);
      return NS_OK;
    }
  }
  return NS_ERROR_FAILURE;  // Invalid component.
}

NS_IMETHODIMP
calRecurrenceRule::SetComponent(const nsACString& aComponentType,
                                nsTArray<int16_t> const& aValues) {
  if (mImmutable) return NS_ERROR_OBJECT_IS_IMMUTABLE;

  // Look up the array for this component type.
  for (int i = 0; recurrenceTable[i].name; ++i) {
    auto const& row = recurrenceTable[i];
    if (aComponentType.EqualsASCII(row.name)) {
      // Found it.
      int16_t* dest = (int16_t*)((uint8_t*)&mIcalRecur + row.offset);
      if (aValues.Length() > row.maxCount) return NS_ERROR_FAILURE;
      for (int16_t v : aValues) {
        *dest++ = v;
      }
      // Terminate array unless full.
      if (aValues.Length() < row.maxCount) {
        *dest++ = ICAL_RECURRENCE_ARRAY_MAX;
      }
      return NS_OK;
    }
  }
  return NS_ERROR_FAILURE;  // Invalid component.
}

/* calIDateTime getNextOccurrence (in calIDateTime aStartTime, in calIDateTime
 * aOccurrenceTime); */
NS_IMETHODIMP
calRecurrenceRule::GetNextOccurrence(calIDateTime* aStartTime,
                                     calIDateTime* aOccurrenceTime,
                                     calIDateTime** _retval) {
  NS_ENSURE_ARG_POINTER(aStartTime);
  NS_ENSURE_ARG_POINTER(aOccurrenceTime);
  NS_ENSURE_ARG_POINTER(_retval);

  nsresult rv;

  nsCOMPtr<calIDateTimeLibical> icaldtstart =
      do_QueryInterface(aStartTime, &rv);
  NS_ENSURE_SUCCESS(rv, rv);

  nsCOMPtr<calIDateTimeLibical> icaloccurtime =
      do_QueryInterface(aOccurrenceTime, &rv);
  NS_ENSURE_SUCCESS(rv, rv);

  struct icaltimetype dtstart;
  icaldtstart->ToIcalTime(&dtstart);

  struct icaltimetype occurtime;
  icaloccurtime->ToIcalTime(&occurtime);

  icalrecur_iterator* recur_iter;
  recur_iter = icalrecur_iterator_new(mIcalRecur, dtstart);
  if (!recur_iter) return NS_ERROR_OUT_OF_MEMORY;

  struct icaltimetype next = icalrecur_iterator_next(recur_iter);
  while (!icaltime_is_null_time(next)) {
    if (icaltime_compare(next, occurtime) > 0) break;

    next = icalrecur_iterator_next(recur_iter);
  }

  icalrecur_iterator_free(recur_iter);

  if (icaltime_is_null_time(next)) {
    *_retval = nullptr;
    return NS_OK;
  }

  nsCOMPtr<calITimezone> tz;
  aStartTime->GetTimezone(getter_AddRefs(tz));
  *_retval = new calDateTime(&next, tz);
  CAL_ENSURE_MEMORY(*_retval);
  NS_ADDREF(*_retval);
  return NS_OK;
}

static inline icaltimetype ensureDateTime(icaltimetype const& icalt) {
  if (!icalt.is_date) {
    return icalt;
  } else {
    icaltimetype ret = icalt;
    ret.is_date = 0;
    ret.hour = 0;
    ret.minute = 0;
    ret.second = 0;
    return ret;
  }
}

NS_IMETHODIMP
calRecurrenceRule::GetOccurrences(calIDateTime* aStartTime,
                                  calIDateTime* aRangeStart,
                                  calIDateTime* aRangeEnd, uint32_t aMaxCount,
                                  nsTArray<RefPtr<calIDateTime>>& aDates) {
  NS_ENSURE_ARG_POINTER(aStartTime);
  NS_ENSURE_ARG_POINTER(aRangeStart);
  aDates.ClearAndRetainStorage();

  // make sure the request is sane; infinite recurrence
  // with no end time is bad times.
  if (!aMaxCount && !aRangeEnd && mIcalRecur.count == 0 &&
      icaltime_is_null_time(mIcalRecur.until))
    return NS_ERROR_INVALID_ARG;

  nsCOMArray<calIDateTime> dates;

#ifdef DEBUG_dbo
  {
    char const* const ss = icalrecurrencetype_as_string(&mIcalRecur);
    nsAutoCString tst, tend;
    aRangeStart->ToString(tst);
    aRangeEnd->ToString(tend);
    printf("RULE: [%s -> %s, %d]: %s\n", tst.get(), tend.get(),
           mIcalRecur.count, ss);
  }
#endif

  nsresult rv;

  nsCOMPtr<calIDateTimeLibical> icalrangestart =
      do_QueryInterface(aRangeStart, &rv);
  NS_ENSURE_SUCCESS(rv, rv);
  nsCOMPtr<calIDateTimeLibical> icaldtstart =
      do_QueryInterface(aStartTime, &rv);
  NS_ENSURE_SUCCESS(rv, rv);

  struct icaltimetype rangestart, dtstart, dtend;
  icalrangestart->ToIcalTime(&rangestart);
  rangestart = ensureDateTime(rangestart);
  icaldtstart->ToIcalTime(&dtstart);
  nsCOMPtr<calITimezone> tz;
  aStartTime->GetTimezone(getter_AddRefs(tz));

  if (aRangeEnd) {
    nsCOMPtr<calIDateTimeLibical> icalrangeend =
        do_QueryInterface(aRangeEnd, &rv);
    NS_ENSURE_SUCCESS(rv, rv);

    icalrangeend->ToIcalTime(&dtend);
    dtend = ensureDateTime(dtend);

    // if the start of the recurrence is past the end,
    // we have no dates
    if (icaltime_compare(dtstart, dtend) >= 0) {
      return NS_OK;
    }
  }

  icalrecur_iterator* recur_iter;
  recur_iter = icalrecur_iterator_new(mIcalRecur, dtstart);
  if (!recur_iter) return NS_ERROR_OUT_OF_MEMORY;

  for (icaltimetype next = icalrecur_iterator_next(recur_iter);
       !icaltime_is_null_time(next);
       next = icalrecur_iterator_next(recur_iter)) {
    icaltimetype const dtNext(ensureDateTime(next));

    // if this thing is before the range start
    if (icaltime_compare(dtNext, rangestart) < 0) {
      continue;
    }

    if (aRangeEnd && icaltime_compare(dtNext, dtend) >= 0) break;

    calIDateTime* cdt = new calDateTime(&next, tz);
    aDates.AppendElement(cdt);
#ifdef DEBUG_dbo
    {
      nsAutoCString str;
      cdt->ToString(str);
      printf("  occ: %s\n", str.get());
    }
#endif
    if (aMaxCount && aMaxCount <= aDates.Length()) break;
  }

  icalrecur_iterator_free(recur_iter);

  return NS_OK;
}

/**
 ** ical property getting/setting
 **/
NS_IMETHODIMP
calRecurrenceRule::GetIcalProperty(calIIcalProperty** prop) {
  icalproperty* const rrule = icalproperty_new_rrule(mIcalRecur);
  CAL_ENSURE_MEMORY(rrule);
  *prop = new calIcalProperty(rrule, nullptr);
  if (!*prop) {
    icalproperty_free(rrule);
    return NS_ERROR_FAILURE;
  }

  NS_ADDREF(*prop);
  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::SetIcalProperty(calIIcalProperty* aProp) {
  NS_ENSURE_ARG_POINTER(aProp);
  nsresult rv;

  nsCOMPtr<calIIcalPropertyLibical> icalprop = do_QueryInterface(aProp, &rv);
  NS_ENSURE_SUCCESS(rv, rv);

  if (mImmutable) return NS_ERROR_OBJECT_IS_IMMUTABLE;

  nsAutoCString propname;
  rv = aProp->GetPropertyName(propname);
  NS_ENSURE_SUCCESS(rv, rv);
  if (propname.EqualsLiteral("RRULE")) {
    mIsNegative = false;
  } else {
    return NS_ERROR_INVALID_ARG;
  }

  icalproperty* prop;
  struct icalrecurrencetype icalrecur;

  prop = icalprop->GetLibicalProperty();

  icalrecur = icalproperty_get_rrule(prop);

  // XXX Note that we ignore the dtstart and use the one from the
  // event, though I realize now that we shouldn't.  Ignoring
  // dtstart makes it impossible to have multiple RRULEs on one
  // event that start at different times (e.g. every day starting on
  // jan 1 for 2 weeks, every other day starting on feb 1 for 2
  // weeks).  Neither the server nor the UI supports this now,
  // but we really ought to!
  // struct icaltimetype icaldtstart;
  // icaldtstrat = icalproperty_get_dtstart(prop);

  if (icalrecur.count != 0)
    mIsByCount = true;
  else
    mIsByCount = false;

  mIcalRecur = icalrecur;

  return NS_OK;
}

NS_IMETHODIMP
calRecurrenceRule::SetIcalString(const nsACString& str) {
  if (mImmutable) return NS_ERROR_OBJECT_IS_IMMUTABLE;

  nsresult rv = NS_OK;
  nsAutoCString name;
  nsCOMPtr<calIICSService> icsSvc = cal::getICSService();
  nsCOMPtr<calIIcalProperty> prop;

  rv = icsSvc->CreateIcalPropertyFromString(str, getter_AddRefs(prop));
  NS_ENSURE_SUCCESS(rv, rv);

  rv = prop->GetPropertyName(name);
  NS_ENSURE_SUCCESS(rv, rv);

  if (!name.EqualsLiteral("RRULE")) {
    return NS_ERROR_ILLEGAL_VALUE;
  }

  return SetIcalProperty(prop);
}

NS_IMETHODIMP
calRecurrenceRule::GetIcalString(nsACString& str) {
  nsresult rv = NS_OK;

  nsCOMPtr<calIIcalProperty> prop;

  rv = this->GetIcalProperty(getter_AddRefs(prop));

  if (NS_SUCCEEDED(rv)) {
    rv = prop->GetIcalString(str);
  }

  return rv;
}

NS_IMETHODIMP
calRecurrenceRule::GetWeekStart(short*) { return NS_ERROR_NOT_IMPLEMENTED; }

NS_IMETHODIMP
calRecurrenceRule::SetWeekStart(short) { return NS_ERROR_NOT_IMPLEMENTED; }
